### ISSUE TITLE
Rancher RKE2 Setup - Cannot make this work yet

### ISSUE THREAD

**Issue by inzanez** [democratic-csi/democratic-csi#188]
Hi

I just tried setting up the deomcratic-csi driver on an RKE2 cluster. It seems that with a CIS 1.6 profile during setup, the cluster usually forbids certain things like privilege use, host network etc.
That can be corrected (I guess?) by binding an unrestricted cluster role to the service accounts running the deomcratic-csi. Most of the errors have cleared up that way, but I am left with this:

`Error: failed to generate container "292265f52e1406f77c0a8b5970247a3729b00a0e00b0c26b8915bc823cc4112d" spec: failed to generate spec: path "/" is mounted on "/" but it is not a shared mount`

The values.yaml used looks as this:
```
csiDriver:
  name: "org.democratic-csi.nfs"
storageClasses:
- name: truenas-nfs-api
  defaultClass: false
  reclaimPolicy: Retain
  volumeBindingMode: Immediate
  allowVolumeExpansion: true
  parameters:
    fsType: nfs
  mountOptions: 
  - noatime
  - nfsvers=3
  secrets:
    provisioner-secret:
    controller-publish-secret:
    node-stage-secret:
    node-publish-secret:
    controller-expand-secret:
driver:
  config:
    driver: freenas-api-nfs
    instance_id:
    httpConnection:
      protocol: https
      host: 172.xxx.xxx.yy
      port: 443
      apiKey: xxxx
      allowInsecure: true
      apiVersion: 2
    zfs:
      datasetParentName: data/k8s/nfs-test/vols
      detachedSnapshotsDatasetParentName: data/k8s/nfs-test/snaps
      datasetEnableQuotas: true
      datasetEnableReservation: false
      datasetPermissionsMode: "0777"
      datasetPermissionsUser: 0
      datasetPermissionsGroup: 0
    nfs:
      shareHost: 172.xxx.xxx.yy
      shareAlldirs: false
      shareAllowedHosts: []
      shareAllowedNetworks: []
      shareMaprootUser: root
      shareMaprootGroup: root
      shareMapallUser: ""
      shareMapallGroup: ""

```

**Comment by inzanez** [comment]
Oh, ...so it seems that for non systemd based host systems (like Alpine for instance) you would have to set  'rshared' mount propagation manually on '/':
`mount --make-rshared /`

So setting that on every worker node corrected the error. Would be interesting why that is required by the democratic-csi driver?...maybe there is a way around that?

**Comment by mattmattox** [comment]
I'm running TrueNAS in NFS & iSCSI mode in my RKE2 clusters. I did need to add PSP role binding to give the pods the permissions it needs.

```
---
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: system-unrestricted-psp-role-control
  namespace: democratic-csi
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: system-unrestricted-psp-role
subjects:
- kind: ServiceAccount
  name: zfs-iscsi-democratic-csi-controller-sa
  namespace: democratic-csi
- kind: ServiceAccount
  name: zfs-nfs-democratic-csi-controller-sa
  namespace: democratic-csi
---
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: system-unrestricted-psp-role-default
  namespace: democratic-csi
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: system-unrestricted-psp-role
subjects:
- kind: ServiceAccount
  name: default
  namespace: democratic-csi
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: system-unrestricted-psp-role-node
  namespace: democratic-csi
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: system-unrestricted-psp-role
subjects:
- kind: ServiceAccount
  name: zfs-iscsi-democratic-csi-node-sa
  namespace: democratic-csi
- kind: ServiceAccount
  name: zfs-nfs-democratic-csi-node-sa
  namespace: democratic-csi
```

**Comment by travisghansen** [comment]
Are those additional permissions/etc something we should get added into the chart directly?

**Comment by inzanez** [comment]
> Are those additional permissions/etc something we should get added into the chart directly?

Not sure,…I guess this is very specific to RKE, RKE2 and the security settings applied to the Kubernetes environment.

Maybe it would make sense to add it to the documentation though?

**Comment by travisghansen** [comment]
Do those actually correlate to PSPs? Those are going away right?

**Comment by inzanez** [comment]
> Do those actually correlate to PSPs? Those are going away right?

Yes, one part is the PSP required (clusterrole allow use of unrestricted psp), the other part is the OS specific mount rshared for non system-d base operating systems.

I guess it would be safe adding a PSP and adding it to the clusterrole for the demicratic-csi,…or just making use of an existing one if there is a general ‘unrestricted’ psp on every kube cluster.

**Comment by inzanez** [comment]
There seems to be another issue. While it seems that I can now create volumes for NFS and iSCSI, iSCSI volumes cannot be mounted:

`MountVolume.MountDevice failed for volume "pvc-b212b409-58fd-49e1-9dda-bf8900398c57" : rpc error: code = Internal desc = {"code":127,"stdout":"","stderr":"env: can't execute 'iscsiadm': No such file or directory\n","timeout":false}`

I have verified that all worker nodes do have 'iscsiadm' available and working. It's at /usr/sbin/iscsiadm. I am not sure where that message comes from though...

**Comment by inzanez** [comment]
Ok, so it seems that the `chroot` command only seeks in /sbin, /bin, /usr/bin, while my binary is located at `/usr/sbin`. A symbolic link resolved the issue. Maybe that path can be added in a future release...

**Comment by travisghansen** [comment]
Sure enough. Not sure how I missed that one but I’ll get it corrected!

**Comment by mattmattox** [comment]
> > Do those actually correlate to PSPs? Those are going away right?
> 
> Yes, one part is the PSP required (clusterrole allow use of unrestricted psp), the other part is the OS specific mount rshared for non system-d base operating systems.
> 
> I guess it would be safe adding a PSP and adding it to the clusterrole for the demicratic-csi,…or just making use of an existing one if there is a general ‘unrestricted’ psp on every kube cluster.

Most applications bring their own PSP as part of the deployment because you never know if someone has changed the default PSPs. So to address this, I created a [PR](https://github.com/democratic-csi/charts/pull/27) for the chart.

**Comment by travisghansen** [comment]
`chroot` issue fixed in this commit: https://github.com/democratic-csi/democratic-csi/commit/81a42d789f9b2c8cdf5fd93e40d740b8a2ae3699

Released in `v1.7.0`

Chart with `psp` is also now available.
